text/0000-macro-metavar-expr.md

```diff
@@ -0,0 +1,332 @@
+with the correct capacity.  To do this, we need to know the number of
+repetitions.
+
+[guide-vec]: https://doc.rust-lang.org/book/ch19-06-macros.html#declarative-macros-with-macro_rules-for-general-metaprogramming
+
+Metadata about metavariables, like the number of repetitions, can be accessed
+using **metavariable expressions**.  The metavariable expression for the
+count of the number of repetitions of a metavariable `x` is `${count(x)}`, so
+we can improve the `vec` macro as follows:
+
+```
+#[macro_export]
+macro_rules! vec {
+    ( $( $x:expr ),* ) => {
+        {
+            let mut temp_vec = Vec::with_capacity(${count(x)});
+            $(
+                temp_vec.push($x);
+            )*
+            temp_vec
+        }
+    };
+}
+```
+
+The following metavariable expressions are available:
+
+| Expression                 | Meaning    |
+|----------------------------|------------|
+| `${count(ident)}`          | The number of times `$ident` repeats in total. |
+| `${count(ident, depth)}`   | The number of times `$ident` repeats at up to `depth` nested repetition depths. |
+| `${index()}`               | The current index of the inner-most repetition. |
+| `${index(depth)}`          | The current index of the nested repetition at `depth` steps out. |
+| `${length()}`              | The length of the inner-most repetition. |
+| `${length(depth)}`         | The length of the nested repetition at `depth` steps out. |
+| `${ignore(ident)}`         | Binds `$ident` for repetition, but expands to nothing. |
+| `$$`                       | Expands to a single `$`, for removing ambiguity in recursive macro definitions. |
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Metavariable expressions in declarative macros provide expansions for
+information about metavariables that are otherwise not easily obtainable.
+
+This is a backwards-compatible change as both `$$` and `${ .. }` are not
+currently accepted as valid.
+
+The metavariable expressions added in this RFC are concerned with declarative
+macro metavariable repetitions, and obtaining the information that the
+compiler knows about the repetitions that are being processed.
+
+## Count
+
+The `${count(x)}` metavariable expression shown in the `vec` example in the
+previous section counts the number of repetitions that will occur if the
+identifier is used in a repetition at this depth.  This means that in a macro
+expansion like:
+
+```
+    ${count(x)} $( $x )*
+```
+
+the expression `${count(x)}` will expand to the number of times the `$( $x )*`
+repetition will repeat.
+
+If repetitions are nested, then an optional depth parameter can be used to
+limit the number of nested repetitions that are counted.  For example, a macro
+expansion like:
+
+```
+    ${count(x, 1)} ${count(x, 2)} ${count(x, 3)} $( a $( b $( $x )* )* )*
+```
+
+The three values this expands to are the number of outer-most repetitions (the
```

Comment:
This might be my ignorance of macros speaking, but does the RFC need to specify which kind of fragment they produce?  Are they allowed to expand to `(0+4)` or `1+1+1+1` or `2*2` or `4_usize`, or only to exactly `4`?  Is there any way I can start to depend on that expansion, like writing a macro that checks a length by only accepting a particular token?

Reply:
It doesn't, but it would be a good change to add it.  It should expand to a literal with the appropriate value and no suffix (i.e. only and exactly `4`).  This allows consistent use in things like `stringify!` and tuple indexing.  Type inferencing should be able to infer the correct type for when it is used in code (and also produce an error if the value that is produced won't fit inside the target type).